server/src/internalClusterTest/java/org/elasticsearch/readiness/ReadinessClusterIT.java

```diff
@@ -0,0 +1,264 @@
+        assertTrue(getStatus(internalCluster().getInstance(ReadinessService.class, masterNode)));
+
+        Integer masterPort = internalCluster().getInstance(ReadinessService.class, internalCluster().getMasterName())
+            .boundAddress()
+            .publishAddress()
+            .getPort();
+
+        assertThat(
+            internalCluster().masterClient()
+                .admin()
+                .cluster()
+                .prepareState()
+                .execute()
+                .actionGet()
+                .getState()
+                .nodes()
+                .getMasterNode()
+                .getName(),
+            equalTo(masterNode)
+        );
+
+        logger.info("--> stop master node");
+        Settings masterDataPathSettings = internalCluster().dataPathSettings(internalCluster().getMasterName());
+        internalCluster().stopCurrentMasterNode();
+
+        assertFalse(getStatus(masterPort));
+
+        try {
+            assertThat(
+                client().admin()
+                    .cluster()
+                    .prepareState()
+                    .setMasterNodeTimeout("100ms")
+                    .execute()
+                    .actionGet()
+                    .getState()
+                    .nodes()
+                    .getMasterNodeId(),
+                nullValue()
+            );
+            fail("should not be able to find master");
+        } catch (MasterNotDiscoveredException e) {
+            // all is well, no master elected
+        }
+
+        logger.info("--> start previous master node again");
+        final String nextMasterEligibleNodeName = internalCluster().startNode(
+            Settings.builder().put(nonDataNode(masterNode())).put(masterDataPathSettings)
+        );
+        assertThat(
+            internalCluster().nonMasterClient()
+                .admin()
+                .cluster()
+                .prepareState()
+                .execute()
+                .actionGet()
+                .getState()
+                .nodes()
+                .getMasterNode()
+                .getName(),
+            equalTo(nextMasterEligibleNodeName)
+        );
+        assertThat(
+            internalCluster().masterClient()
+                .admin()
+                .cluster()
+                .prepareState()
+                .execute()
+                .actionGet()
+                .getState()
+                .nodes()
+                .getMasterNode()
+                .getName(),
+            equalTo(nextMasterEligibleNodeName)
+        );
+    }
+
+    public void testReadinessDuringRestartsNormalOrder() throws Exception {
+        internalCluster().setBootstrapMasterNodeIndex(0);
+        logger.info("--> start master node");
+        String masterNode = internalCluster().startMasterOnlyNode();
+        internalCluster().validateClusterFormed();
+
+        assertThat(
+            internalCluster().masterClient()
```

Comment:
nit: this is a long chain repeated several times. Perhaps have a helper method to make each individual test easier to read?